app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -550,22 +557,93 @@ protected void deliverError(CommCareSetupActivity receiver,
+    /**
+     * Scan SMS messages for texts with profile references.
+     * @param installTriggeredManually if scan was triggered manually, then
+     *                                 install automatically if reference is found
+     */
+    private void performSMSInstall(boolean installTriggeredManually){
+        String profileLink = null;
+        try {
+            profileLink = this.scanSMSLinks();
+        } catch(SignatureException e){
+            // possibly we want to do something more severe here? could be malicious
+            e.printStackTrace();
+            Toast.makeText(this, Localization.get("menu.sms.not.verified"), Toast.LENGTH_LONG).show();
+        } catch(IOException e){
+            // couldn't get the acutal message payload - possible internet issue?
+            e.printStackTrace();
+            Toast.makeText(this, Localization.get("menu.sms.not.retrieved"), Toast.LENGTH_LONG).show();
+        }
+        if (profileLink != null) {
+            // we found a valid profile link, either start install automatically
+            // or move to READY_TO_INSTALL state
+            Log.v("install", "Performing SMS install with link : " + incomingRef);
+            incomingRef = profileLink;
+            if (installTriggeredManually) {
+                startResourceInstall();
+            } else {
+                uiState = UiState.READY_TO_INSTALL;
+                Toast.makeText(this, Localization.get("menu.sms.ready"), Toast.LENGTH_LONG).show();
+            }
+        } else if(installTriggeredManually) {
+            // only notify if this was manually triggered, since most people won't use this
+            Toast.makeText(this, Localization.get("menu.sms.not.found"), Toast.LENGTH_LONG).show();
+        }
+    }
+
+
+
+    /**
+     * Scan the SMS inbox, looking for messages that meet the expected install format,
+     * and if found and verified return the discovered install link. Current behavior will search
+     * backwards from the most recent text, returning the first discovered valid link
+     * @return the verified install link, null if none found
+     * @throws SignatureException if we discovered a valid-looking message but could not verifyMessageSignatureHelper it
+     */
+    private String scanSMSLinks() throws SignatureException, IOException{
+        // http://stackoverflow.com/questions/11301046/search-sms-inbox
+        final Uri SMS_INBOX = Uri.parse("content://sms/inbox");
+        Cursor cursor = getContentResolver().query(SMS_INBOX, null, null, null, "date desc");
+        try {
+            if (cursor.moveToFirst()) { // must check the result to prevent exception
+                while (!cursor.isAfterLast()) {
+                    String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
+                    if (textMessageBody.contains(GlobalConstants.SMS_INSTALL_KEY_STRING)) {
+                        String installLink = SigningUtil.retrieveParseAndVerifyURL(textMessageBody);

```

Comment:
`retrieveParseAndVerifyURL` expects only the url, so you should parse out the url here with substring (the text message is `[commcare app - do not delete] <url>`